### PR TITLE
feat: sto details and cancel sto procedure

### PR DESCRIPTION
**Pull request by shuffledex**
https://polymath.atlassian.net/browse/MSDK-426
https://polymath.atlassian.net/browse/MSDK-443

### CODE REVIEW

File: src/testUtils/mocks/entities.ts [hunk 1]
```diff
@@ -5,6 +5,7 @@ import { merge } from 'lodash';
 import sinon, { SinonStub } from 'sinon';
 
 import { ProposalDetails, ProposalStage /*, ProposalState */ } from '~/api/entities/Proposal/types';
+import { StoDetails } from '~/api/entities/Sto/types';
```
Comment (monitz87) [inline src/testUtils/mocks/entities.ts:8]: Import this from `~/types`

File: src/utils/conversion.ts [hunk 2]
```diff
@@ -63,6 +66,7 @@ import {
   VenueType as MeshVenueType,
 } from 'polymesh-types/types';
 
+import { FundraiserStatus, StoDetails, Tier } from '~/api/entities/Sto/types';
```
Comment (monitz87) [inline src/utils/conversion.ts:69]: Import these from `~/types`

File: src/api/entities/Sto/types.ts [hunk 3]
```diff
@@ -1,3 +1,32 @@
+import BigNumber from 'bignumber.js';
+
+import { DefaultPortfolio, Identity, NumberedPortfolio, SecurityToken, Venue } from '~/internal';
 export interface Placeholder {
   foo: number;
 }
+
+export interface Tier {
+  total: BigNumber;
```
Comment (monitz87) [inline src/api/entities/Sto/types.ts:9]: amount

File: src/utils/__tests__/conversion.ts [hunk 4]
```diff
@@ -40,13 +40,15 @@ import {
 } from 'polymesh-types/types';
 import sinon from 'sinon';
 
+import { FundraiserStatus } from '~/api/entities/Sto/types';
```
Comment (monitz87) [inline src/utils/__tests__/conversion.ts:43]: Import from `~/types`

File: src/api/entities/Sto/types.ts [hunk 5]
```diff
@@ -1,3 +1,32 @@
...
+  Frozen = 'Frozen',
+  Closed = 'Closed',
+}
+
+export interface StoDetails {
+  creator: Identity;
+  offeringPortfolio: NumberedPortfolio | DefaultPortfolio;
+  offeringAsset: SecurityToken;
+  raisingPortfolio: NumberedPortfolio | DefaultPortfolio;
+  raisingAsset: SecurityToken;
```
Comment (monitz87) [inline src/api/entities/Sto/types.ts:25]: Should be  `raisingCurrency: string`

File: src/utils/conversion.ts [hunk 6]
```diff
@@ -2244,3 +2249,65 @@ export function middlewarePortfolioToPortfolio(
...
+  return {
+    creator: new Identity({ did: identityIdToString(creator) }, context),
+    offeringPortfolio: meshPortfolioIdToPortfolio(offeringPortfolio, context),
+    offeringAsset: new SecurityToken({ ticker: tickerToString(offeringAsset) }, context),
+    raisingPortfolio: meshPortfolioIdToPortfolio(raisingPortfolio, context),
+    raisingAsset: new SecurityToken({ ticker: tickerToString(raisingAsset) }, context),
+    tiers: tiers.map(tier => fundraiserTierToTier(tier)),
+    venue: new Venue({ id: u64ToBigNumber(venueId) }, context),
+    start: momentToDate(start),
+    end: !end.isEmpty ? momentToDate(end.unwrap()) : undefined,
```
Comment (monitz87) [inline src/utils/conversion.ts:2309]: You can use `isSome` instead of negating `isEmpty` (should be `isNone` instead of `isEmpty` in any case)

**Review by monitz87 — CHANGES_REQUESTED** [review]

**Review by monitz87 — CHANGES_REQUESTED** [review]
You forgot to create the `close` method in the `Sto` entity

File: src/api/procedures/__tests__/cancelOffering.ts [hunk 7]
```diff
@@ -0,0 +1,113 @@
+import BigNumber from 'bignumber.js';
+import sinon from 'sinon';
+
+import { FundraiserStatus } from '~/api/entities/Sto/types';
```
[the remark this answers is not in the document]
Comment (shuffledex) [reply src/api/procedures/__tests__/cancelOffering.ts:4]: do you mean move FundraiserStatus to ~/types?

File: src/testUtils/mocks/entities.ts (hunk 1 above)
Comment (shuffledex) [reply src/testUtils/mocks/entities.ts:8]: the same question here. Should I define StoDetails in ~/types instead /entities/Sto/types? Or I have to import from /entities/Sto/types?

File: src/api/entities/Sto/types.ts [hunk 8]
```diff
@@ -1,32 +1,34 @@
...
-  Closed = 'Closed',
-}
-
 export interface StoDetails {
   creator: Identity;
   offeringPortfolio: NumberedPortfolio | DefaultPortfolio;
   offeringAsset: SecurityToken;
   raisingPortfolio: NumberedPortfolio | DefaultPortfolio;
-  raisingAsset: SecurityToken;
+  raisingCurrency: SecurityToken;
```
Comment (monitz87) [inline src/api/entities/Sto/types.ts:22]: This should be a string

File: src/api/entities/Sto/types.ts [hunk 9]
```diff
@@ -1,32 +1,34 @@
...
-export interface Placeholder {
-  foo: number;
+
+export enum StoStatus {
+  Live = 'Live',
+  Frozen = 'Frozen',
+  Closed = 'Closed',
 }
 
 export interface Tier {
```
Comment (monitz87) [inline src/api/entities/Sto/types.ts:11]: Have this extend StoTier

**Review by monitz87 — CHANGES_REQUESTED** [review]

**Review by monitz87 — APPROVED** [review]

**Pull request merged**